### ISSUE TITLE
Writes inserted into memtable may be interpreted using incorrect schema version on schema alters of a counter table

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#5095]
Scylla version (or git commit hash): 2.3+

Introduced in 27014a23d76ecbb169f75b73c2cf4f3b885e7178.

Relevant code:
```
void partition_entry::apply(const schema& s, const mutation_partition& mp, const schema& mp_schema)
{
    apply(s, mutation_partition(s, mp), mp_schema);
}
```

`mutation_partition(s, mp)` is incorrect, because it uses `s` to interpret `mp`, while it should use `mp_schema`.

We may hit this if the current node has a newer schema than the incoming mutation. This can happen during alter when we receive the mutation from a node which hasn't processed the schema change yet.

This is undefined behavior in general. If the alter was adding or removing columns, this may result in corruption of the write where values of one column are inserted into a different column.

**Comment by tgrabiec** [comment]
I re-inspected all the code paths which lead to this, and it seems that this affects only counter tables.

**Comment by avikivity** [comment]
Issue is fixed on master and branch-3.1, but my backport failed on next-3.0 (and probably 2.3).

@tgrabiec please provide a backport for 3.0.

**Comment by tgrabiec** [comment]
@avikivity `git cherry-pick -x -m 1 83bc59a` applies cleanly on next-3.0 here. Or did you mean something else by "failed"?

**Comment by slivne** [comment]
@tgrabiec / @avikivity ping

**Comment by tgrabiec** [comment]
I pushed to next-3.0, let's see what happens

**Comment by avikivity** [comment]
I meant that it did not apply cleanly for me. Maybe I made a mistake, or maybe something else was merged meanwhile.

**Comment by avikivity** [comment]
The tests failed, so something was indeed wrong.

**Comment by slivne** [comment]
this is merged into 3.3, 3.2, 3.1
